data/posts/fog-of-war.md

```diff
@@ -0,0 +1,181 @@
+
+You can take this one step further in Remix via [`<Link prefetch>`][link-prefetch], which allows you to prefetch the route data and components before a user even clicks a link. That way, when the link is clicked the navigation can be instantaneous:
+
+<img alt="Prefetching network diagram" src="/blog-images/posts/fog-of-war/prefetch.png" class="m-auto w-4/5 border rounded-md shadow" />
+
+## The Route Manifest
+
+Let's define a few important terms for clarity:
+
+- **Route tree**: A tree of routes which define the URLs your app can match via parent/child relationships
+- **Route definition**: Aspects of the route used to match a URL - `path`, `index`, `children`
+- **Route implementation**: Aspects of a route needed to load data and render the UI (`loader`, `Component`, `ErrorBoundary`, etc.)
+
+In order to implement these above optimizations, Remix needs to know all of your route _definitions_ in the client so that it can match routes based on nothing more than a `<Link to>`. Once a link is clicked and routes are matched, Remix can issue data fetches and download the route _implementations_ in parallel.
+
+To do this, Remix ships a **Route Manifest** to the client that contains all of your route definitions plus a small amount of metadata. This manifest allows Remix to build a client-side route tree without including the underlying route implementations. Those implementations are loaded via [route.lazy][route-lazy] when a route is navigated to.
+
+Over time, as the user navigates around - more and more route implementations are downloaded via `route.lazy` and the route tree becomes more and more complete.
+
+This approach works for most applications - since the manifest is pretty lightweight and compresses well since it's a repetitive key/value JSON structure. For example, the manifest for https://remix.run/ contains 50 routes, weighs 19.6Kb, but only sends 2.6Kb over the wire after compression.
+
+However, Remix doesn't only want to to provide good performance for small-to-medium sized applications - we want _large_ and _extremely large_ applications to be fast by default too!
+
+Thankfully, we love dog-fooding React Router and Remix on the myriad of applications at Shopify, both internal and public-facing. As we began rolling out Remix to https://www.shopify.com we realized _just how big_ that site is. When you take into account all of the routes and their internationalized URLs (i.e., `/pricing`, `/en/pricing`, `/es/precios` and many more) - the app had over 1300 routes! And because Remix doesn't have a good solution for URL aliasing (yet!), many of the route entries were duplicates pointing to the same route module - and thus duplicating the module information (it's path, it's other module `imports`, etc.). This resulted in a manifest that was over 10Mb uncompressed, and ~85Kb compressed. On slower devices, this could have a noticeable impact on page load times for the device to decompress/parse/compile/execute the JS module.
```

Comment:
Take it or leave it, but "thankfully" might sound a bit too pat-ourselves-on-the-back. I was trying to consider a more take-responsibility approach, but happy for you to go whatever direction you like.

```suggestion
Because we love dog-fooding React Router and Remix on the myriad of applications at Shopify, both internal and public-facing, we knew our current strategy was not cutting it for those larger applications. As we began rolling out Remix to https://www.shopify.com we realized _just how big_ that site is. When you take into account all of the routes and their internationalized URLs (i.e., `/pricing`, `/en/pricing`, `/es/precios` and many more) - the app had over 1300 routes! And because Remix doesn't have a good solution for URL aliasing (yet!), many of the route entries were duplicates pointing to the same route module - and thus duplicating the module information (it's path, it's other module `imports`, etc.). This resulted in a manifest that was over 10Mb uncompressed, and ~85Kb compressed. On slower devices, this could have a noticeable impact on page load times for the device to decompress/parse/compile/execute the JS module.
```